lib/puma/cluster.rb

```diff
@@ -37,7 +37,23 @@ def stop_workers
       @workers.each { |x| x.term }
 
       begin
-        @workers.each { |w| Process.waitpid(w.pid) }
+        t_st = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+        pids = @workers.map(&:pid)
+        until pids.empty?
+          wait = false
+          pids.each do |w_pid|
+            if Process.waitpid(w_pid, Process::WNOHANG)
+              log "    worker status: #{$?}"
+              pids.delete w_pid
+              break
+            else
+              wait = true
+            end
+          end
+          sleep 0.5 if wait
+        end
```

Comment:
There are two cycles instead of one.

You can use `redo`.

See https://github.com/puma/puma/pull/1738#discussion_r264748021

Reply:
@AlexWayfer

Thanks.  As mentioned, I did it this way so it would loop thru all the workers on first pass, then go back for the nil returns.  I'd rather use `Process.waitpid(-1, Process::WNOHANG)`, as that's a bit cleaner, but it seems to repeatedly fail on trunk JIT, I never tried with 2.6.x JIT...

Reply:
> As mentioned, I did it this way so it would loop thru all the workers on first pass, then go back for the nil returns.

It's strange algorithm.

Option 1, with `redo`:

1. Go through all workers
2. Sleep and redo on the first fail (wait every first fail)

Option 2:

1. Go through all workers and attempt to stop them
2. Delete from array in success, wait after all if any failed
3. Retry if there are any remained

Option 3, your choice, as I see:

1. Go through all workers
2. Delete in success, and retry (go further)
3. Don't do anything in fail, just sleep in the end
4. Retry until there are any

I think, the first option is simplest (but may be longest), the second is most optimal, and the third is strange.

Reply:
I changed the logic to use reject!

Reply:
> I changed the logic to use reject!

You can use `loop do` instead of `while true do`. 😉

Reply:
I often look here:
https://msp-greg.github.io/ruby_trunk/file.control_expressions.html

`loop do` isn't mentioned there.  Time to review kernel methods again...